cachito/workers/pkg_managers/gomod.py

```diff
@@ -304,52 +281,149 @@ def resolve_gomod(app_source_path, request, dep_replacements=None, git_dir_path=
+                    _, dep_version = _get_name_and_version(dep.module)
 
-                pkg_level_deps.append({"name": dep_name, "type": "go-package", "version": version})
+                pkg_deps.append({"type": "go-package", "name": dep_name, "version": dep_version})
 
-            # Top-level packages always use the module version
-            pkg = {"name": pkg_name, "type": "go-package", "version": module_version}
-            packages.append({"pkg": pkg, "pkg_deps": pkg_level_deps})
+            main_pkg = {
+                "type": "go-package",
+                "name": pkg.import_path,
+                "version": main_module_version,
+            }
+            main_packages.append({"pkg": main_pkg, "pkg_deps": pkg_deps})
 
-        allowlist = _get_allowed_local_deps(module_name)
-        log.debug("Allowed local dependencies for %s: %s", module_name, allowlist)
-        _vet_local_deps(module_level_deps, module_name, allowlist, app_source_path, git_dir_path)
-        for pkg in packages:
+        allowlist = _get_allowed_local_deps(main_module_name)
+        log.debug("Allowed local dependencies for %s: %s", main_module_name, allowlist)
+        _vet_local_deps(
+            main_module_deps, main_module_name, allowlist, app_source_path, git_dir_path
+        )
+        for pkg in main_packages:
             # Local dependencies are always relative to the main module, even for subpackages
-            _vet_local_deps(pkg["pkg_deps"], module_name, allowlist, app_source_path, git_dir_path)
-            _set_full_local_dep_relpaths(pkg["pkg_deps"], module_level_deps)
+            _vet_local_deps(
+                pkg["pkg_deps"], main_module_name, allowlist, app_source_path, git_dir_path
+            )
+            _set_full_local_dep_relpaths(pkg["pkg_deps"], main_module_deps)
+
+        return {
+            "module": main_module,
+            "module_deps": main_module_deps,
+            "packages": main_packages,
+        }
 
-        return {"module": module, "module_deps": module_level_deps, "packages": packages}
+
+def _deduplicate_to_gomod_dicts(
+    modules: Iterable[GoModule], user_specified_deps_to_replace: set[str]
+) -> list[dict[str, Any]]:
+    modules_by_name_and_version: dict[tuple[str, str], dict[str, Any]] = {}
+    for mod in modules:
+        name, version = _get_name_and_version(mod)
+        # get the module dict for this name+version or create a new one
+        gomodule = modules_by_name_and_version.setdefault(
+            (name, version),
+            {"type": "gomod", "name": name, "version": version, "replaces": None},
+        )
+        # report user-specified replacements (note that those must replace to a version, not a path)
+        if mod.replace and mod.replace.version and mod.path in user_specified_deps_to_replace:
+            gomodule["replaces"] = {"type": "gomod", "name": mod.path, "version": mod.version}
+    return [module for _, module in sorted(modules_by_name_and_version.items())]
```

Comment:
Compared to cachi2, this is extended with "replaces" handling. The tricky part is that replacements are only reported in `go list -deps -json`, not in `go mod download -json`

go mod download -json
```json
{
    "Path": "github.com/cachito-testing/retrodep/v2",
    "Version": "v2.1.1",
    "Info": "{gomodcache_dir}/cache/download/github.com/cachito-testing/retrodep/v2/@v/v2.1.1.info",
    "GoMod": "{gomodcache_dir}/cache/download/github.com/cachito-testing/retrodep/v2/@v/v2.1.1.mod",
    "Zip": "{gomodcache_dir}/cache/download/github.com/cachito-testing/retrodep/v2/@v/v2.1.1.zip",
    "Dir": "{gomodcache_dir}/github.com/cachito-testing/retrodep/v2@v2.1.1",
    "Sum": "h1:inziEuX1Zo/BJv1pgZ2tRPltfQVJaVpRmIlsomaKMb0=",
    "GoModSum": "h1:1ktnevkSAzwQww4E/ICy2dlIouSNdvozRVgq6Umkt/E="
}
```
go list -deps -json all
```json
{
    "ImportPath": "github.com/release-engineering/retrodep/v2/retrodep/glide",
    "Module": {
        "Path": "github.com/release-engineering/retrodep/v2",
        "Version": "v2.1.0",
        "Replace": {
            "Path": "github.com/cachito-testing/retrodep/v2",
            "Version": "v2.1.1",
            "Time": "2019-06-19T12:05:35Z",
            "Dir": "{gomodcache_dir}/github.com/cachito-testing/retrodep/v2@v2.1.1",
            "GoMod": "{gomodcache_dir}/cache/download/github.com/cachito-testing/retrodep/v2/@v/v2.1.1.mod"
        },
        "Dir": "{gomodcache_dir}/github.com/cachito-testing/retrodep/v2@v2.1.1",
        "GoMod": "{gomodcache_dir}/cache/download/github.com/cachito-testing/retrodep/v2/@v/v2.1.1.mod"
    }
}
```
This algorithm deduplicates the modules by their post-replacement name+version, and makes sure to set the `"replaces"` field if the module is reported as a replacement in go list -deps -json

Reply:
Nice :+1: 
I'll steal this for Cachi2 (as we also need to dedup go list vs go download outputs) :laughing: